plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx

```diff
@@ -109,3 +120,79 @@ export const addProductCollectionBlockToParentOfPaginationBlock = () => {
 		);
 	}
 };
+
+export const useSetPreviewState = ( {
+	setPreviewState,
+	location,
+	attributes,
+	setAttributes,
+}: {
+	setPreviewState?: SetPreviewState | undefined;
+	location: WooCommerceBlockLocation;
+	attributes: ProductCollectionAttributes;
+	setAttributes: (
+		attributes: Partial< ProductCollectionAttributes >
+	) => void;
+} ) => {
+	const setState = ( newPreviewState: PreviewState ) => {
+		setAttributes( {
+			__privatePreviewState: {
+				...attributes.__privatePreviewState,
+				...newPreviewState,
+			},
+		} );
+	};
+
+	// Running setPreviewState function provided by Collection, if it exists.
+	useLayoutEffect( () => {
+		if ( ! setPreviewState ) {
+			return;
+		}
+
+		const cleanup = setPreviewState?.( {
+			setState,
+			location,
+			attributes,
+		} );
+
+		if ( cleanup ) {
+			return cleanup;
+		}
+
+		// It should re-run only when setPreviewState changes to avoid performance issues.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ setPreviewState ] );
+
+	/**
+	 * For all Product Collection blocks that inherit query from the template,
+	 * we want to show a preview message in the editor if the block is in
+	 * generic archive template i.e.
+	 * - Products by category
+	 * - Products by tag
+	 * - Products by attribute
+	 */
+	useLayoutEffect( () => {
+		if ( ! setPreviewState ) {
+			const isGenericArchiveTemplate =
+				location.type === LocationType.Archive &&
+				location.sourceData?.termId === null;
+			if ( isGenericArchiveTemplate ) {
+				setAttributes( {
+					__privatePreviewState: {
+						isPreview: !! attributes?.query?.inherit,
+						previewMessage: __(
+							'Actual products will vary depending on the page being viewed.',
+							'woocommerce'
+						),
+					},
+				} );
+			}
+		}
+	}, [
+		attributes?.query?.inherit,
+		location.sourceData?.termId,
```

Comment:
We'd probably need the `location.sourceData?.productId` in the deps here to work properly when switching between specific-single-product-templates

Reply:
It will re-run whenever we switch between the templates because the Component itself will re-render. I suggest we test this scenario while exposing `registerProductCollection` in follow-up PR. If it doesn't re-trigger, we can definitely add this dependency. Does that sound good to you?

Reply:
Yup! Since the final form of this helper will include the `uses_context`, we can reiterate!